### DISCUSSION TITLE
issue + feature request: defaultValue of type `Date` is not applied to`<input type=date>` - with `register` API

### DISCUSSION THREAD

**Discussion by tresorama** [react-hook-form/react-hook-form#11860] [Q&A]
## What

Doing this...
```tsx
const Component = () => {
  const { register } = useForm({
    defaultValue: {
      day: new Date("2024-12-31"),
    }
  })

  return <input type="date" {...register('day', {valueAsDate: true })} />
}
```
... result in this
<img width="113" alt="Schermata 2024-04-22 alle 18 43 27" src="https://github.com/react-hook-form/react-hook-form/assets/47954700/42e68794-8829-41a7-988e-a90dbe9c4ce5">
an "input" in  blank/empty state, instead of showing an already selected date.  
This means that the defaultValue defined in `useForm` is not respected.

Even if we define `defaultValue` as input props the outcome is the same
```tsx
  return <input type="date" defaultValue="2024-12-31" {...register('day', {valueAsDate: true })} />
```
<img width="113" alt="Schermata 2024-04-22 alle 18 43 27" src="https://github.com/react-hook-form/react-hook-form/assets/47954700/42e68794-8829-41a7-988e-a90dbe9c4ce5">

### Version Number

7.51.3

### Codesandbox/Expo snack

https://stackblitz.com/edit/vitejs-vite-v95kgv?file=src%2Fsections%2Foriginal-issue.tsx

### Steps to reproduce

1. Open sandbox
2. See the pink form field (key:`day`) that looks as "empty" even if default value is present
3. Look also that `watch` show that default value is memory state is correct

### Note 1
It seems that the culprit is the `ref` returned from `register` that is some way overwrite `defaultValue`.  
Look for `day3` input where i omitted passing `ref` and it show the default value correctly.  
Because there is no `ref`, it's not a valid workaround. react hooks form can no longer track field value changes.

### Note 2
A workaround is to use `useEffect` and update the value of the field after first render.  
Look for `day2` input where I did this.  It show the default value correctly.  
But it's "hacky" and inconvenient.  

### Expected behaviour

This code should let the browser input field show that a date is already selected.
Now it show a blank/empty state.
```jsx
<input 
  type="date" 
  defaultValue="2024-04-04"
  {...register('day', {valueAsDate:true})}
/>
```

Tested only in Chrome.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by tresorama** [comment]
Even if would be great to have a solution for this, **like a standard API for transform/cast between react-hookk-form state and html widget state...** 

---

In this [sandbox](https://stackblitz.com/edit/vitejs-vite-v95kgv?file=src%2Fsections%2Ffix-1.tsx) I tried to overcome this with a restructure of data flow.  
The idea is to split the schema into 2 parts: `formSchema` and `dataSchema`.  
```tsx
const dataSchema = z.object({ ... }) // schema that rapresent how datasource fetched from DB looks

const formSchema = z
  .object({ ... }) // validate react-hook-form state
  .transform((input)=> // then cast types for dataSchema validation)
  .pipe(dataSchema); // Validate against the dataSchema
```
dataSchema is the schema used for validating at the Datasource (DB) level.  


So the validation schema has three step
1. Validate react-hook-form formValues
2. If pass, cast types to expected type of the Datasource layer
3. Validate against the dataSchema 

This is more verbose, but more declarative and easily readable

**Comment by tresorama** [comment]
To give better context to this issue...

## Goal
Let's say the you want a form with an input the let you select a date, but you want to set a default date.  
Doing that when the user open the form a date is already in place (for example the "today" date)

## This does not work
```tsx
const Component = () => {
  const { register } = useForm({
    defaultValue: {
      day: new Date("2024-12-31"),
    }
  })

  return <input type="date" {...register('day', {valueAsDate: true })} />
}
```
Output:
<img width="113" alt="Schermata 2024-04-22 alle 18 43 27" src="https://github.com/react-hook-form/react-hook-form/assets/47954700/42e68794-8829-41a7-988e-a90dbe9c4ce5">



## This works - use Controller API
If you instead switch from `register` API to `<Controller>` API you can:
- edit how *reach-hook-form => set `value` to => html input element*
- edit how *react-hook-form => get `value` from the `onChange` event and set it to => react-hook-form state*

```tsx
const Component = () => {
  const { register } = useForm({
    defaultValue: {
      day: new Date("2024-12-31"),
    }
  })

  return (
  <Controller
    control={control}
    name="day"
    render={({field}) => (
      <input
        {...field}
        type="date"
        value={field.value ? field.value.toISOString().substring(0,10) : undefined}
        onChange={( {target:{value}} ) => field.onChange( value === '' ? undefined : new Date(value) ) }
      />
    )}
  />
  )
}
```
#### Consideration
But this means that you cannot choose which api to use (register or Controller), but as soon as you need a date input you MUST use the controller api, which is less performant

#### Open Question
Why not exposed in the registr api a similar way to manipulate the value between react-hook-form state and html input state (DOM state) ??
Something like
```tsx
<input 
  type="date" 
  {...register("day", {
    // input value => react-hook-form state
    getValueFromInputAs:  value => value === '' ? undefined : new Date(value), 
    // react-hook-form-state => input value
    setValueToInputAs: value => value ? value.toISOString().substring(0,10) : undefined, 
  })}
/>
```


## This works - use string instead of `Date` and accept a complex validation schema
Because this doesn't work ...
```tsx
const schema = z.object({
  day: z.date(),
});
const Component = () => {
  const { register } = useForm<z.input<typeof schema>>({
    defaultValue: {
      day: new Date("2024-12-31"),
    }
  })
  // ...
}
```
... you can switch the field type from `Date` to `string` like ...
```tsx
const { register } = useForm<z.input<typeof schema>>({
    defaultValue: {
      day: "2024-12-31", // now it's a string
    }
  })
```
... but the validation schema will complain (showing Typescript errors).  You can fix this by adding more complexity to the schema, with downside of coupling the schema with HTML input behavior.
```tsx
const schema = z.object({
  // 1. validate html input values
  day: z.string().min(1, "Required field"),
}).transform(input => ({
 // 2. If step 1 is valid, cast/coerce data types for the step 3
  day: new Date(input.day),
}).pipe(z.object({
  // 3. validate data
  day: z.date(),
});
```

... and updating also `useForm` generic types to have correct inference inside `handleSubmit`
```tsx
type SchemaInput = z.input<typeof schema>;
type SchemaOutput = z.output<typeof schema>;

const { register, handleSubmit } = useForm<SchemaInput, any, SchemaOutput>({
    defaultValue: {
      day: "2024-12-31", // now it's a string
    }
  })

<form onSubmit={handleSubmit((formValues) => {
  // here formValues has SchemaOutput types and not SchemaInput
  // so "formValues.day" is of type Date, not string 
}}

**Comment by tresorama** [comment]
## Feature request

Why not exposeing in the `register` api a way to manipulate the value between react-hook-form state and DOM html input state ??
This will give "parity" with the Controller API.

Something like
```tsx
<input 
  type="date" 
  {...register("day", {
    // input value => react-hook-form state
    getValueFromInputAs:  value => value === '' ? undefined : new Date(value), 
    // react-hook-form-state => input value
    setValueToInputAs: value => value ? value.toISOString().substring(0,10) : undefined, 
  })}
/>
```

**Comment by kujohn** [comment]
I am having this issue as well. `defaultValue` doesn't work at all with `type=date` inputs.

**Comment by laurent512** [comment]
same problem with date indeed

**Comment by dennemark** [comment]
I like this idea.
Normal html requires <[input type="date" defaultValue="YYYY-MM-DD"](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/date)> and this is how it is stored on the input.value. There is another Date stored on valueAsDate and react-hook-form seems to use this. In my case this leads to a weird behaviour, when the input element can be disposed within the form component. On first render the default value is available, on second render react-hook-form tries to set a Date value onto defaultValue which is wrong. 
It would be very convenient to just set a Date onto defaultValue and react-hook-form converts it to 'YYYY-MM-DD'

related: https://github.com/react-hook-form/react-hook-form/issues/10883
